File: runtime/java/units/wasi/fd_tell.java
// requires: memory/i64_store
int wasi_fd_tell(int fd, int outPtr) {
    Object e = fds.get(fd);
    if (isStdio(e)) {
        return WASI_SPIPE;
    }
    if (!(e instanceof Handle)) {
        return WASI_BADF;
    }
    try {
        memory.i64_store(Integer.toUnsignedLong(outPtr), ((Handle) e).ch.position());
    } catch (java.io.IOException ex) {
        return WASI_IO;
    }
    return WASI_OK;
}

File: runtime/go/units/wasi/fd_tell.go
// requires: memory/i64_store
func (w *WASI) wasi_fd_tell(fd, outPtr uint32) uint32 {
    f, ok := w.fds[fd].(*os.File)
    if !ok {
        return wasiBadf
    }
    if w.isStdio(f) {
        return wasiSpipe
    }
    pos, err := f.Seek(0, 1) // whence 1 = SEEK_CUR
    if err != nil {
        return wasiIo
    }
    w.memory.i64_store(uint64(outPtr), uint64(pos))
    return wasiOk
}
